015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,133 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Proxy-Based Kafka Per-Topic Encryption
+
+The goal of this proposal is to provide topic-level encryption-at-rest for Kafka such that distinct encryption keys can be used for different topics. 
+A core library is proposed which is deployed into a proxy. Proxies can flexibly be deployed in a variety of topologies and require no changes to Kafka clients and brokers. All encryption/decryption takes place at the proxy. An unencrypted topic's messages pass through the proxy without being changed and without
+paying any performance penalty.
+
+This proposal is based on a complete, working implementation.
+
+
+## Current situation
+
+Strimzi does not have a current capability in this area. 
+
+## Motivation
+
+Apache Kafka does not directly support any form of encryption-at-rest for data stored at a broker.
+Nevertheless, Kafka is increasingly used as a store of data, not just as a
+means of transferring it from one location to another. In an
+enterprise, this means that Kafka must conform to the same security and
+compliance requirements as conventional data storage systems such as relational databases. 
+To compensate, Kafka service providers typically use disk or file system encryption. 
+This approach has well-known shortcomings, most notably the ability of anyone with appropriate file system permissions, such as system administrators, to read data.  Security best practices and compliance standards
+explicitly call for application- or database-level encryption to protect sensitive information against such exposures.
+This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.
+
+
+## Proposal
+
+An implementation of topic encryption is proposed whereby the message stream between client and broker is
+intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload
+should be encrypted according to a policy.  If so, the data portions are encrypted and the modified
```

Comment:
what "data portions" means here? just the kafka message payload? what if sensible information are transferred using Kafka headers in the message?

Reply:
By "data portions" we are referring to the map of records, keyed by topic, in a Kafka Produce and Fetch messages which are inspected. We have not considered encrypting record headers but this is an interesting suggestion which could be easily accommodated.

As the entire message is navigated to the record level, it would be natural apply encryption to record headers as well. This would require a change to the encryption policy to indicate header names or topic/header combinations to encrypt. The logic change to support header encryption would be minimal.

Reply:
fyi @SeanRooooney @zrlvep